HSBM.py

```diff
@@ -0,0 +1,442 @@
+
+    if embed_kws:
+        if not isinstance(embed_kws, dict):
+            raise TypeError("embed_kws must be a dict")
+
+    if not isinstance(cluster_kws, dict):
+        raise TypeError("cluster_kws must be a dict")
+
+
+class RecursiveCluster(NodeMixin, BaseEstimator):
+    def __init__(
+        self,
+        selection_criteria=None,
+        cluster_method="GMM",
+        n_init=1,
+        parent=None,
+        min_components=1,
+        max_components=10,
+        cluster_kws={},
+        min_split=1,
+        max_level=20,
+        delta_criter=None,
+        likelihood_ratio=None,
+    ):
+
+        _check_common_inputs(
+            min_components, max_components, cluster_kws, embed_kws=None
+        )
+
+        if cluster_method not in ["GMM", "KMeans", "Spherical-KMeans"]:
+            msg = "clustering method must be one of {GMM, Kmeans, Spherical-KMeans}"
+            raise ValueError(msg)
+
+        if delta_criter:
+            if delta_criter < 0:
+                raise ValueError("delta_criter must be positive")
+
+        if likelihood_ratio:
+            if likelihood_ratio <= 0 or likelihood_ratio >= 1:
+                raise ValueError("likelihood_ratio must be in (0,1)")
+
+        self.parent = parent
+        self.n_init = n_init
+        self.min_components = min_components
+        self.max_components = max_components
+        self.cluster_method = cluster_method
+        self.min_split = min_split
+        self.selection_criteria = selection_criteria
+        self.cluster_kws = cluster_kws
+        self.min_split = min_split
+        self.max_level = max_level
+        self.delta_criter = delta_criter
+        self.likelihood_ratio = likelihood_ratio
+        
+    def fit(self, X, adj=None):
+        self.fit_predict(X)
+        return self
+
+    def predict(self, X, adj=None):
+        check_is_fitted(self, ["model_"], all_or_any=all)
+        labels = self.fit_predict(X)
+        return labels
+    
+    def fit_predict(self, X, level=None, adj=None):
+        X = check_array(
+            X, dtype=[np.float64, np.float32],  ensure_min_samples=1
+        )
+        if level:
+            if not isinstance(level, int):
+                raise TypeError("level must be an int")
+
+        self.X = X
+        self.adj = adj
+
+        if self.max_components > X.shape[0]:
+            msg = "max_components must be >= n_samples, but max_components = "
+            msg += "{}, n_samples = {}".format(self.max_components, X.shape[0])
+            raise ValueError(msg)
+        
+        while True:
+            current_node = self._get_next_node()
+            if current_node:
+                current_node._cluster_and_decide()
+            else:
+                break
+
+        labels = self._to_labels()
+        if level:
+            labels = labels[:, level]
+
+        self.predictions = labels
+        return labels
+
+    def _get_next_node(self):
+        leaves = self.root.leaves
```

Comment:
it seems odd to me that this is `root.leaves` and not just `self.leaves` but maybe I don't understand the recursion... can you explain?

Reply:
So the way I imagine it should work is this: go through all leaves in the current tree and pick the first node that satisfy some conditions which are 1) not too few samples to be fitted w/ AGMM or not fewer samples than min_split; 2) its depth not greater than max_level; 3) not yet fitted. I think `root.leaves` makes sure that it go through all the leaves in the overall tree. But I can also see how to iterate w/ `self.leaves`. Once a node gets `k=1`, the algorithm checks its siblings or moves up and check some ancestor's siblings. What do you prefer?